source/extensions/filters/http/cache/cache_filter.h

```diff
@@ -80,21 +82,97 @@ class CacheFilter : public Http::PassThroughFilter,
   // Http::StreamEncoderFilter
   Http::FilterHeadersStatus encodeHeaders(Http::ResponseHeaderMap& headers,
                                           bool end_stream) override;
-  Http::FilterDataStatus encodeData(Buffer::Instance& buffer, bool end_stream) override;
-  Http::FilterTrailersStatus encodeTrailers(Http::ResponseTrailerMap& trailers) override;
 
   static LookupStatus resolveLookupStatus(absl::optional<CacheEntryStatus> cache_entry_status,
                                           FilterState filter_state);
 
 private:
+  class UpstreamRequest : public Http::AsyncClient::StreamCallbacks, public InsertQueueCallbacks {
+  public:
+    void sendHeaders(Http::RequestHeaderMap& request_headers);
+    void disconnectFilter();
+
+    // StreamCallbacks
+    void onHeaders(Http::ResponseHeaderMapPtr&& headers, bool end_stream) override;
+    void onData(Buffer::Instance& data, bool end_stream) override;
+    void onTrailers(Http::ResponseTrailerMapPtr&& trailers) override;
+    void onComplete() override;
+    void onReset() override;
+
+    // InsertQueueCallbacks
+    void insertQueueOverHighWatermark() override;
+    void insertQueueUnderLowWatermark() override;
+    void insertQueueAborted() override;
+
+    static UpstreamRequest* create(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
+                                   Http::AsyncClient& async_client,
+                                   const Http::AsyncClient::StreamOptions& options);
+    UpstreamRequest(CacheFilter* filter, std::shared_ptr<HttpCache> cache,
```

Comment:
That's annoying with smart pointers, unfortunately. (Yeah you can do a `unique_ptr` wrapped around a `new` instead of a `make_unique`, I think that's more problematic than a public constructor in what is essentially a private context anyway.)